### ISSUE TITLE
[Bug]: TypeError: Cannot read property 'type' of null

### ISSUE THREAD

**Issue by liby** [babel/babel#13636]
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

babel-loader (webpack)

### Input code

The issue was found in a company private project and there is no way to give the source code. If I finish the minimal reproduce later, I will definitely post it here.

### Configuration file name

babel.config.js

### Configuration

<details>
<summary>babel.config.js</summary>

```js
module.exports = function (api) {
  api.cache(true);

  return {
    presets: [
      [
        "@babel/preset-env",
        {
          useBuiltIns: "usage",
          corejs: 3,
          modules: "commonjs",
        },
      ],
      [
        "@babel/preset-react",
        {
          runtime: "automatic",
        },
      ],
      [
        "@babel/preset-typescript",
        {
          isTSX: true,
          allExtensions: true,
        },
      ],
    ],
    plugins: [
      "babel-plugin-styled-components",
      "babel-plugin-const-enum",
      ["@babel/plugin-proposal-decorators", { legacy: true }],
      ["@babel/plugin-proposal-private-methods", { loose: true }],
      ["@babel/plugin-proposal-class-properties", { loose: true }],
      ["@babel/plugin-proposal-private-property-in-object", { loose: true }],
      "@babel/proposal-object-rest-spread",
      "@babel/plugin-transform-runtime",
      "@babel/plugin-proposal-optional-chaining",
    ],
  };
};
```
</details>

### Current and expected behavior

-  Expected behavior: compiled successfully

- Current behavior:
    <details>
    <summary>Full Log</summary>
    
    ```
    ❯ yarn run build:dev
    yarn run v1.22.10
    $ env-cmd -f .env.dev --use-shell "yarn run build:webpack"
    $ webpack --color --mode production --config scripts/webpack/webpack.build.js
    tsconfig-paths-webpack-plugin: Using config file at /Project/tsconfig.json
    assets by status 36.7 MiB [cached] 279 assets
    orphan modules 1.03 MiB [orphan] 330 modules
    runtime modules 8.33 KiB 33 modules
    javascript modules 22.3 MiB
      modules by path ./src/application/ 5.93 MiB 1503 modules
      modules by path ./node_modules/ 16.3 MiB 1327 modules
    asset modules 3.77 KiB (javascript) 1.12 MiB (asset)
      modules by path ./node_modules/@private/config/dist/assets/logo/images/*.svg 2.99 KiB (javascript) 599 KiB (asset) 73 modules
      modules by path ./node_modules/@private/config/dist/assets/logo/images/*.png 714 bytes (javascript) 245 KiB (asset) 17 modules
      modules by path ./node_modules/@private/config/dist/assets/logo/images/*.jpg 84 bytes (javascript) 301 KiB (asset) 2 modules
    css modules 1.67 MiB
      modules by path ./src/application/client/ 933 KiB 5 modules
      3 modules
    json modules 24.4 KiB
      ./src/application/shared/config/application-config.json 50 bytes [built] [code generated]
      ./node_modules/@private/config/dist/user/user.config.json 24.3 KiB [built] [code generated]
    
    ERROR in ./src/application/shared/hooks/use-sort-by.ts
    Module build failed (from ./node_modules/babel-loader/lib/index.js):
    TypeError: Cannot read property 'type' of null
        at Object.isAssignable (/Project/node_modules/@babel/core/node_modules/@babel/parser/lib/index.js:9605:18)
        at /Project/node_modules/@babel/core/node_modules/@babel/parser/lib/index.js:10470:52
        at Array.every (<anonymous>)
        at Object.isAssignable (/Project/node_modules/@babel/core/node_modules/@babel/parser/lib/index.js:10470:30)
        at Object.isAssignable (/Project/node_modules/@babel/core/node_modules/@babel/parser/lib/index.js:9613:22)
        at /Project/node_modules/@babel/core/node_modules/@babel/parser/lib/index.js:9788:40
        at Array.every (<anonymous>)
        at Object.shouldParseArrow (/Project/node_modules/@babel/core/node_modules/@babel/parser/lib/index.js:9788:21)
        at Object.parseParenAndDistinguishExpression (/Project/node_modules/@babel/core/node_modules/@babel/parser/lib/index.js:11799:28)
        at Object.parseExprAtom (/Project/node_modules/@babel/core/node_modules/@babel/parser/lib/index.js:11419:23)
     @ ./src/application/admin/screens/user/user.component.tsx 48:17-53
     @ ./src/application/admin/router.component.tsx 22:42-96
     @ ./src/application/admin/app.component.tsx 32:37-66
     @ ./src/application/admin/index.tsx 9:34-60
    
    webpack 5.48.0 compiled with 1 error in 80241 ms
    error Command failed with exit code 1.
    info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
    error Command failed with exit code 1.
    info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
    ```
    </details>

### Environment

  ```
  System:
    OS: macOS 10.15.7
  Binaries:
    Node: 12.20.1 - ~/.nvs/default/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 6.14.10 - ~/.nvs/default/bin/npm
  npmPackages:
    @babel/cli: ^7.14.8 => 7.14.8
    @babel/core: ^7.15.0 => 7.15.0
    @babel/plugin-proposal-decorators: ^7.14.5 => 7.14.5
    @babel/plugin-proposal-object-rest-spread: ^7.14.7 => 7.14.7
    @babel/plugin-proposal-optional-chaining: ^7.14.5 => 7.14.5
    @babel/plugin-transform-runtime: ^7.15.0 => 7.15.0
    @babel/preset-env: ^7.15.0 => 7.15.0
    @babel/preset-react: ^7.14.5 => 7.14.5
    @babel/preset-typescript: ^7.15.0 => 7.15.0
    @babel/runtime: ^7.14.8 => 7.14.8
    @babel/runtime-corejs3: ^7.14.9 => 7.14.9
    babel-loader: ^8.2.2 => 8.2.2
    babel-plugin-const-enum: ^1.1.0 => 1.1.0
    babel-plugin-styled-components: ^1.13.2 => 1.13.2
    webpack: ^5.48.0 => 5.48.0
```

**Comment by nicolo-ribaudo** [comment]
Does this only happen since the v7.15.0 release (10 hours ago)?

**Comment by liby** [comment]
> Does this only happen since the v7.15.0 release (10 hours ago)?
Right. Downgrade to the following version and the error disappears.
```
❯ npx envinfo --preset babel

  System:
    OS: macOS 10.15.7
  Binaries:
    Node: 12.20.1 - ~/.nvs/default/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 6.14.10 - ~/.nvs/default/bin/npm
  npmPackages:
    @babel/cli: ^7.14.8 => 7.14.8
    @babel/core: ^7.14.8 => 7.14.8
    @babel/plugin-proposal-decorators: ^7.14.5 => 7.14.5
    @babel/plugin-proposal-object-rest-spread: ^7.14.7 => 7.14.7
    @babel/plugin-proposal-optional-chaining: ^7.14.5 => 7.14.5
    @babel/plugin-transform-runtime: ^7.14.5 => 7.14.5
    @babel/preset-env: ^7.14.9 => 7.14.9
    @babel/preset-react: ^7.14.5 => 7.14.5
    @babel/preset-typescript: ^7.14.5 => 7.14.5
    @babel/runtime: ^7.14.8 => 7.14.8
    @babel/runtime-corejs3: ^7.14.9 => 7.14.9
    babel-loader: ^8.2.2 => 8.2.2
    babel-plugin-const-enum: ^1.1.0 => 1.1.0
    babel-plugin-styled-components: ^1.13.2 => 1.13.2
    webpack: ^5.48.0 => 5.48.0
```

**Comment by nicolo-ribaudo** [comment]
Ok, the bug is probably when parsing an arrow function or an expression in parentheses in `./src/application/shared/hooks/use-sort-by.ts`, if that helps creating a small example when reproducing the bug. It's _likely_ that the closing `)` that triggers the bug is then followed by `:`.

**Comment by liby** [comment]
@nicolo-ribaudo
<details>
<summary>use-sort-by.ts</summary>

```
import { useCallback, useEffect, useMemo, useState } from "react";
import _, { ListIteratee } from "lodash";

/**
 * SortOrder is an enumeration of the possible sort orderings.
 *
 * Nameing ref:
 * Microrsoft https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sortorder?view=netframework-4.8
 * JDK https://docs.oracle.com/javase/7/docs/api/javax/swing/SortOrder.html
 */
export const enum SortOrder {
  // Enumeration value indicating the items are unordered.
  Unspecified = "unspecified",
  // Enumeration value indicating the items are sorted in increasing order.
  Ascending = "asc",
  // Enumeration value indicating the items are sorted in decreasing order.
  Descending = "desc",
}

export interface ISortRule<T> {
  columnName: string;
  orderBy?: ListIteratee<T>;
  orders?: SortOrder[];
}

export type ICurrentOrders = Record<string, SortOrder>;
export type ISortFunction = (columnName: string) => void;

export interface ISortByHookResult<T> {
  sortedData: T[];
  currentOrders: ICurrentOrders;
  sort: ISortFunction;
}

/**
 *
 * @param data Original Data
 * @param sortRules sort rules
 * @returns {
 *  sortedData: the sorted data, with the same interface as data
 *  currentOrders: {
 *    "columnA": SortOrder.Unspecified
 *    "columnB": SortOrder.Ascending
 *  }
 *  sort: ISortFunction
 * }
 */
export const useSortBy = <T = unknown>(
  data: T[],
  sortRules: ISortRule<T>[],
  tick = 0
): ISortByHookResult<T> => {
  const initialSortOrders = useMemo(() => {
    return _.fromPairs(
      sortRules.map((sortRule) => [sortRule.columnName, SortOrder.Unspecified])
    );
  }, [sortRules]);

  const [currentOrders, setCurrentOrders] =
    useState<ICurrentOrders>(initialSortOrders);
  const [sortedData, setSortedData] = useState<T[]>(data);

  const sort: ISortFunction = useCallback(
    (columnName: string) => {
      const sortRule = sortRules.find(
        (eachSortRule) => eachSortRule.columnName === columnName
      );
      const orders = sortRule.orders || [
        SortOrder.Unspecified,
        SortOrder.Ascending,
        SortOrder.Descending,
      ];

      // get next order of this column
      const currentOrder = currentOrders[columnName];
      const currentOrderIndex = orders.findIndex(
        (order) => order === currentOrder
      );
      const nextOrderIndex =
        currentOrderIndex === orders.length - 1 ? 0 : currentOrderIndex + 1;
      const nextOrder = orders[nextOrderIndex];

      // sort data by the rule
      if (nextOrder === SortOrder.Unspecified) {
        setSortedData(data);
      } else {
        setSortedData(
          _.orderBy(
            data,
            [sortRule.orderBy || sortRule.columnName],
            [nextOrder]
          )
        );
      }
      // reset other column to unspecified and change the order of this column
      setCurrentOrders({
        ...initialSortOrders,
        [columnName]: nextOrder,
      });
    },
    // Add tick here to forceUpdate for mutable data
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [currentOrders, data, initialSortOrders, sortRules]
  );

  // Resort when tick update. to force update the sorted data when
  // data is mutable data
  useEffect(() => {
    if (tick > 0) {
      const currentOrder = Object.entries(currentOrders).find(
        ([, order]): boolean => order !== SortOrder.Unspecified
      );

      if (currentOrder) {
        const [columnName, sortOrder] = currentOrder;
        const sortRule = sortRules.find(
          (eachSortRule) => eachSortRule.columnName === columnName
        );

        const newSortedData = _.orderBy(
          data,
          [sortRule.orderBy || sortRule.columnName],
          [sortOrder as SortOrder.Ascending | SortOrder.Descending]
        );
        setSortedData((prevSortedData) =>
          _.isEqual(prevSortedData, newSortedData)
            ? prevSortedData
            : newSortedData
        );
      } else {
        setSortedData(data);
      }
    }
  }, [currentOrders, data, sortRules, tick]);

  return useMemo(() => {
    return {
      currentOrders,
      sortedData,
      sort,
    };
  }, [currentOrders, sortedData, sort]);
};
```
</details>

**Comment by nicolo-ribaudo** [comment]
Thanks! A minimal example is `([, a]): b => c`. We'll fix this soon in the next patch.

**Comment by nicolo-ribaudo** [comment]
Released in `@babel/parser@7.15.2`
